### PR TITLE
Fix recursive whitelisting and handling of opaque for derive default

### PR DESCRIPTION
**Pull request by photoszzt**
Fix regression of derive default analysis.   
Also fix opaque handing exposed by the above fix. r? @fitzgen

### CODE REVIEW

File: src/ir/analysis/derive_default.rs [hunk 1]
```diff
@@ -158,7 +158,7 @@ impl<'ctx, 'gen> MonotoneFramework for CannotDeriveDefault<'ctx, 'gen> {
             }
         };
 
-        if ty.is_opaque(self.ctx, item) {
+        if item.is_opaque(self.ctx, &()) {
```
Comment (fitzgen) [inline src/ir/analysis/derive_default.rs]: Ah, is the idea that we missed the `opaque_by_name` case before?

Comment (photoszzt) [inline src/ir/analysis/derive_default.rs]: Yes.

**Pull request merged**